### ISSUE TITLE
Regulated Community ("public") Editing Rules

### ISSUE THREAD

**Issue by nathankota** [uic-utah/uic-inventory#143]
## General Rules
- [x] To enter a Site a user's profile must be completed
- [x] Regulated Community users can only edit/view items they create
- [x] A draft inventory can only be created once a Site is complete

## Site
- [x] Site pages are editable and Sites can be deleted until a Well Inventory is submitted 
- [x] Once the first Well Inventory is submitted, the Site cannot be edited or deleted

#### If only draft inventories are associated with the site: 
- [x] Site and draft inventory pages are editable
- [x] Site and draft inventory can be deleted

## Inventory Status = `Submitted` and beyond
- [x] Site and inventory cannot be edited or deleted
- [x] Existing draft inventories can be edited or deleted
- [x] New (draft) inventory can be created and submitted

**Comment by nathankota** [comment]
@steveoh are some or all of these rules in place that I should be testing?

**Comment by steveoh** [comment]
It's a bit mish mash but yes you should definitely be trying to break things.

**Comment by nathankota** [comment]
@rsparker-utah please confirm that these rules are correct, and that all necessary rules are captured in this issue

**Comment by rsparker-utah** [comment]
@nathankota We've reviewed the above rules and I can confirm they are correct.

**Comment by nathankota** [comment]
@rsparker-utah please test and provide feedback

**Comment by steveoh** [comment]
I believe these are implemented

**Comment by nathankota** [comment]
Ref: #301

**Comment by steveoh** [comment]
this is ready to be validated

**Comment by rsparker-utah** [comment]
I tested these rules on 5/2/2024 in the morning and they were all working as expected. Planning to do a second round of validation after #316 is resolved.

**Comment by rsparker-utah** [comment]
I tested the following Regulated Community ("public") Editing Rules:

**General Rules**

- To enter a Site a user's profile must be completed - **Validated**.
- Regulated Community users can only edit/view items they create - **Validated**. I tested this by logging into my public test account and was only able to see the sites/inventories that were created using this public test account. 
- A draft inventory can only be created once a Site is complete - **Validated**. I tested this in my public account in two ways: 1.) By creating a new site and completing all of the information except adding a contact that was an owner, owner/operator, or legal representative. The app prevented the site from being created until I added an owner, owner/operator, or legal representative contact. 2.) I also tried creating a site and filling out everything except for the site location. I then went back to the main page and tried creating an inventory. The app prevented me from creating an inventory until the site location information was complete. After I resolved these two checks, I was able to create a draft inventory as expected.

**Site**

- Site pages are editable and Sites can be deleted until a Well Inventory is submitted - **Validated**.
- Once the first Well Inventory is submitted, the Site cannot be edited or deleted - **Validated**. 

I tested the above by deleting a site without any inventories and also sites that only had draft inventories. Once a site had an inventory with a submitted status, I was then unable to delete the site as expected. 

**Inventory Status = Draft**
If only draft inventories are associated with the site:

- Site and draft inventory pages are editable - **Validated**. 
- Site and draft inventory can be deleted - **Validated**.

If there is another inventory in one of the statuses below...

**Inventory Status = Submitted and beyond**

- Site and inventory cannot be edited or deleted - **Validated**. I tested this by trying to edit a site in my public account that had an existing "Submitted" status (and beyond) inventory. I was unable to edit both the site and submitted inventory as expected. 
- Existing draft inventories can be edited or deleted - **Validated**.
- New (draft) inventory can be created and submitted - **Validated**.

This looks good to me and based on my testing is ready to be closed.

**Comment by steveoh** [comment]
> * I tested this by logging into my public test account and was only able to see the sites/inventories that were created using this public test account.

Did you try to edit the url to visit sites, inventories, and admin routes that you shouldn't have access to?

**Comment by rsparker-utah** [comment]
Oh! No I hadn't thought to try that during my original testing. I went back and tried the following:

- I went into my admin account and navigated to the Inventory Review Page of a Submitted inventory (i.e., 267) for site (i.e., 129) that was originally created by my public account. I then copied this Inventory Review Page url (https://uic-inventory.dev.utah.gov/review/site/129/inventory/267) and pasted it into a separate browser logged into my public account. I tried this same process with a few different status inventories across a few different sites that were created by different users.
- I then navigated to the accounts tab in my admin account (https://uic-inventory.dev.utah.gov/admin/accounts), copied the link, and pasted it into a separate browser logged into my public account.

In all of the above cases I was prevented (as expected) from accessing the links and directed to the "How did you get here?" page in my public account, for example:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/22bf46d8-3037-4dab-b247-af6d77ebaf04)

Next using my admin account, I tried navigating to the site details, site contacts, and site location pages of a site (i.e., 124) not created by either my admin account or my public account. I copied the the link for each of these and pasted it into a separate browser logged into my public account. In this case, instead of being directed to the "How did you get here?" page, my public account was routed to the incomplete site notification page.

For example:
For the "Site Details" page link (https://uic-inventory.dev.utah.gov/site/124/add-details):
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/5184d44e-d8fb-4bc1-998b-0a7ef3fa02b1)

For the "Site Contacts" page link (https://uic-inventory.dev.utah.gov/site/124/add-contacts):
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/55dc62d3-3e39-45ab-89d2-859d4ced373d)

For the "Site Location" page link (https://uic-inventory.dev.utah.gov/site/124/add-location):
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/4087ac8c-b470-48f6-a4d3-93d5a6ad8ebc)

Clicking "Back" on any of these directed my public account back to the main "Your sites and inventory" page.

**Comment by rsparker-utah** [comment]
Would it be possible to make it so in those cases for the public account where it was showing the incomplete inventory notification when navigating to links the account shouldn't have access to (i.e., the site details, contacts, and location for sites the account didn't create), for example:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/79314043-d136-488f-8e9a-4d43f5b831d0)

it instead showed the "How did you get here?" page? for example:
![image](https://github.com/uic-utah/uic-inventory/assets/49735267/5da844d6-e2fc-4685-89bf-bdd9098bd03b)

**Comment by nathankota** [comment]
@rsparker-utah it sounds like the editing rules are validated, and perhaps we create a new/different issue for the "How did you get here?"
 page?

**Comment by rsparker-utah** [comment]
@nathankota Yes, that sounds good to me. I created a separate issue (i.e., #322 ) to address the "How did you get here?" page adjustment request.

**Comment by steveoh** [comment]
We'll want to keep testing these scenarios as I make changes for the admin rules to make sure we don't regress.

**Comment by rsparker-utah** [comment]
I completed another round of testing today (5/17/2024) for the public editing rules. Based on my testing, the rules are working as expected.

I tried the following during this round of testing:

### General Rules

- To enter a Site a user's profile must be completed - **Validated**.
- Regulated Community users can only edit/view items they create - **Validated**. I tested this by logging into my public test account and was only able to see the sites/inventories that were created using this public test account. I also logged into my admin account and copied the urls for the site information (i.e., details, contacts, and location pages) and inventory information (i.e., location, well details, and sign/submit pages) for sites that were not created by my public user account (i.e., sites 135 and 149). I then logged into my public user account in a different browser and tried accessing each of these copied urls. For each of these urls, my public user was directed to the “You are not allowed to be here…” screen as expected. 
- A draft inventory can only be created once a Site is complete - **Validated**. I repeated the tests from my last round testing within my public account by: 1.) By creating a new site and completing all of the information except adding a contact that was an owner, owner/operator, or legal representative. The app prevented the site from being created until I added an owner, owner/operator, or legal representative contact. 2.) I also tried creating a site and filling out everything except for the site location. I then went back to the main page and tried creating an inventory. The app prevented me from creating an inventory until the site location information was complete. After I resolved these two checks, I was able to create a draft inventory as expected.

### Site

- Site pages are editable and Sites can be deleted until a Well Inventory is submitted - **Validated**. I tested this by creating a new site and well inventory in my public account. I then tested editing the site and well inventory information while the inventory was in the “Draft” status. As expected, I was able to edit all site and well inventory information while the well inventory status was “Draft”.
- Once the first Well Inventory is submitted, the Site cannot be edited or deleted - **Validated**. I tested this in my public account on a site (i.e., 147) that contained well inventories at each of the post “Draft” well statuses (i.e., Submitted, Under Review, Approved, Authorized, Completed). I then tried to edit the site and submitted well inventory information. As expected, I was not able to edit any of the site information once the well inventory status was past “Draft”. I also attempted deleting the site and was prevented from doing so, as expected.

### Inventory Status = Draft
If only draft inventories are associated with the site:

- Site and draft inventory pages are editable - **Validated**. I tested this by creating a new site and well inventory in my public account. I then tested editing well inventory information while the inventory was in the “Draft” status. As expected, I was able to edit all well inventory information while the well inventory status was “Draft”.
- Site and draft inventory can be deleted - **Validated**. Similarly, I tested this on the same site using my public account. I tried deleting a well with “Draft” status and was able to do so as expected. I also tried deleting a draft inventory on a site that also had other existing inventories with statuses other than “Draft” and was able to delete the “Draft” inventory as expected.

If there is another inventory in one of the statuses below...

### Inventory Status = Submitted and beyond

- Site and inventory cannot be edited or deleted - **Validated**. I tested this by trying to edit a site in my public account that had an existing "Submitted" status (and beyond) inventory. I was unable to edit both the site and submitted inventory as expected.
- Existing draft inventories can be edited or deleted - **Validated**.
- New (draft) inventory can be created and submitted - **Validated**.

**Comment by steveoh** [comment]
can this also be closed?

**Comment by rsparker-utah** [comment]
Yes, thank you
